agent/engine/host_resource_manager.go

```diff
@@ -16,20 +16,313 @@
+		h.consumedResource[MEMORY].SetIntegerValue(mem)
+
+		// PORTS
+		portsResource, ok := resources[PORTSTCP]
+		if ok {
+			h.consumedResource[PORTSTCP].StringSetValue = append(h.consumedResource[PORTSTCP].StringSetValue, portsResource.StringSetValue...)
+		}
+
+		// PORTS_UDP
+		portsResource, ok = resources[PORTSUDP]
+		if ok {
+			h.consumedResource[PORTSUDP].StringSetValue = append(h.consumedResource[PORTSUDP].StringSetValue, portsResource.StringSetValue...)
+		}
+
+		// GPU
+		gpu := *h.consumedResource[GPU].IntegerValue + *resources[GPU].IntegerValue
+		h.consumedResource[GPU].SetIntegerValue(gpu)
+
+		// Set consumed status
+		h.taskConsumed[taskArn] = true
+		logger.Info("Resources successfully consumed, continue to task creation", logger.Fields{"taskArn": taskArn})
+		return true, nil
+	}
+	logger.Info("Resources not consumed, enough resources not available", logger.Fields{"taskArn": taskArn})
+	return false, nil
+}
+
+// Functions checkConsumableIntType and checkConsumableStringSetType to be called
+// only after checking for resource map health
+func (h *HostResourceManager) checkConsumableIntType(resourceName string, resources map[string]*ecs.Resource) bool {
+	resourceConsumableStatus := *(h.initialHostResource[resourceName].IntegerValue) >= *(h.consumedResource[resourceName].IntegerValue)+*(resources[resourceName].IntegerValue)
+	return resourceConsumableStatus
+}
+
+func (h *HostResourceManager) checkConsumableStringSetType(resourceName string, resources map[string]*ecs.Resource) bool {
+	resourceSlice := resources[resourceName].StringSetValue
+
+	// (optimizization) Get a resource specific map to ease look up
+	resourceMap := make(map[string]struct{}, len(resourceSlice))
+	for _, v := range resourceSlice {
+		resourceMap[*v] = struct{}{}
+	}
+
+	// Check intersection of resource StringSetValue is empty with consumedResource
+	for _, obj1 := range h.consumedResource[resourceName].StringSetValue {
+		_, ok := resourceMap[*obj1]
+		if ok {
+			// If resource is already reserved by some other task, this 'resources' object can not be consumed
+			return false
+		}
+	}
+	return true
+}
+
+func checkResourceExistsInt(resourceName string, resources map[string]*ecs.Resource) error {
+	_, ok := resources[resourceName]
+	if ok {
+		if resources[resourceName].IntegerValue == nil {
+			return &ResourceIsNilForTask{resourceName}
+		}
+	} else {
+		return &ResourceNotFoundForTask{resourceName}
+	}
+	return nil
+}
+
+func checkResourceExistsStringSet(resourceName string, resources map[string]*ecs.Resource) error {
+	_, ok := resources[resourceName]
+	if ok {
+		for _, obj := range resources[resourceName].StringSetValue {
+			if obj == nil {
+				return &ResourceIsNilForTask{resourceName}
+			}
+		}
+	} else {
+		return &ResourceNotFoundForTask{resourceName}
+	}
+	return nil
+}
+
+// Checks all resources exists and their values are not nil
+func checkResourcesHealth(resources map[string]*ecs.Resource) error {
```

Comment:
I'm not sure if we necessarily want to make sure that all resource types exist in the map. Technically, for tasks that don't consume certain resource types, it should be okay for them to leave it out. 

We might want to do check, instead:
```
for each key,val in resources:
  does key exist in host resources? i.e. is it a recognized resource type? if not, throw error
  if valid key, then check if value type matches
  Lastly, check if the specific type of value is not nil
```